File: python/popular_domains.py
#!/usr/bin/python

import sys

# Write a script that takes a list of email addresses on STDIN and outputs the 10 
# most common domains (including counts).

domains = {}

for line in sys.stdin:
	line = line.rstrip()

	local, domain = line.split('@')

	if domain not in domains:
		domains[domain] = 1
	else:
		domains[domain] += 1

counter = 0;
for domain, count in sorted(domains.items(), lambda a, b: cmp(a[1], b[1]), reverse=True):
	if counter == 10:
		break
	counter += 1
	print "%r => %r" % (domain, count)

File: perl/popular_domains.pl
#!/usr/bin/perl

use strict;
use warnings;

# Write a script that takes a list of email addresses on STDIN and outputs the 10 
# most common domains (including counts).

my %domains;

# loop the input data from STDIN
while (<STDIN>){
	chomp;
	
	# get the local part and the domain
	my ($local, $domain) = split(/@/, $_);
	$domains{$domain}++;
}

my @sorted_domains = sort { $domains{$b} <=> $domains{$a} } keys %domains;

my $count = 0;
foreach my $domain (@sorted_domains) {
	last if $count == 10;
	$count++;
	printf "%s => %d\n", $domain, $domains{$domain};
}
